_posts/2024-08-06-datafusion-python-udf-comparisons.md

```diff
@@ -0,0 +1,595 @@
+        suppkey = suppkey_arr[idx].as_py()
+        returnflag = returnflag_arr[idx].as_py()
+        value = (partkey, suppkey, returnflag)
+        result.append(value in values_of_interest)
+
+    return pa.array(result)
+
+
+is_of_interest = udf(
+    is_of_interest_impl,
+    [pa.int64(), pa.int64(), pa.utf8()],
+    pa.bool_(),
+    "stable",
+)
+
+df_udf_filter = df_lineitem.filter(
+    is_of_interest(col("l_partkey"), col("l_suppkey"), col("l_returnflag"))
+)
+```
+
+When working with a DataFusion UDF in python, you define your function to take in some number of
+expressions. During the evaluation, these will get computed into their corresponding values and
+passed to your UDF as a pyarrow Array. We must return an Array also with the same number of
+elements (rows). So the UDF example just iterates through all of the arrays and checks to see if
+the tuple created from these columns matches any of those that we’re looking for.
+
+I’ll repeat because this is something that tripped me up the first time I wrote a UDF for
+datafusion: **Datafusion UDFs, even scalar UDFs, process an array of values at a time not a single
+row.** This is different from some other dataframe libraries and you may need to recognize a slight
+change in mentality.
+
+Some important lines here are the lines like `partkey = partkey.as_py()`. When we do this, we pay a
+heavy cost. Now instead of keeping the analysis in the rust code, we have to take the values in the
+array and convert them over to python objects. In this case we end up getting two numbers and a
+string as real python objects, complete with reference counting and all. Also we are iterating
+through the array in python rather than rust native. These will **significantly** slow down your
+code. Any time you have to cross the barrier where you change values inside the rust arrays into
+python objects or vice versa you will pay **heavy** cost in that transformation. You will want to
+design your UDFs to avoid this as much as possible.
+
+## Python approach using pyarrow compute
+
+DataFusion uses [Apache Arrow](https://arrow.apache.org/) as it’s in memory data format. This can
+be seen in the way that Arrow Arrays are passed into the UDFs. We can take advantage of the fact
+that [pyarrow](https://github.com/apache/arrow-rs), the python arrow wrapper, provides a variety of
+useful functions. In the example below, we are only using a few of the boolean functions and the
+equality function. Each of these functions takes two arrays and analyzes them row by row. In the
+below example, we shift the logic around a little since we are now operating on an entire array of
+values instead of checking a single row ourselves.
+
+```python
+import pyarrow.compute as pc
+
+def udf_using_pyarrow_compute_impl(
+    partkey_arr: pa.Array,
+    suppkey_arr: pa.Array,
+    returnflag_arr: pa.Array,
+) -> pa.Array:
+    results = None
+    for partkey, suppkey, returnflag in values_of_interest:
```

Comment:
I think you can use `pyarrow.is_in` to speed this up, instead of doing an equality check multiple times: https://arrow.apache.org/docs/python/generated/pyarrow.compute.is_in.html